instance/models/openedx_instance.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Open edX Instance models
+"""
+from django.conf import settings
+from django.db import models
+
+from instance.gandi import GandiAPI
+from instance.logging import log_exception
+from .mixins.openedx_database import OpenEdXDatabasesMixin
+from .mixins.version_control import GitHubInstanceMixin
+from .instance import Instance
+from .openedx_appserver import OpenEdXAppConfiguration, OpenEdXAppServer
+
+# Constants ###################################################################
+
+gandi = GandiAPI()
+
+# Models ######################################################################
+
+
+class BaseOpenEdXInstance(Instance, OpenEdXAppConfiguration, OpenEdXDatabasesMixin):
+    """
+    BaseOpenEdXInstance: represents a website or set of affiliated websites powered by the same
+    OpenEdX installation.
+    """
+
+    # Most settings/fields are inherited from OpenEdXAppConfiguration
+    sub_domain = models.CharField(max_length=50)
+    base_domain = models.CharField(max_length=50, blank=True)
+
+    active_appserver = models.OneToOneField(OpenEdXAppServer, null=True, on_delete=models.SET_NULL, related_name='+')
```

Comment:
@bradenmacdonald 

>> Why not query this dynamically based on the appserver statuses, rather than a OneToOne field?

> That's easy: Because I wanted this to enforce that only one server is active at a time, and this accomplishes that very cleanly. The other approach, storing a boolean "active" field on the OpenEdXAppServer model and using unique constraints and carefully managing transactions to change status (e.g. "set active=false where id=1; set active=true where id=2" in a transaction), is more complex and harder to get right.

That makes sense, I see. (And that was what I meant, yes, not that the change should be automatic - "dynamic" was for the fact that instead of storing an AppServer id in the Instance model field, the `active_appserver` attribute would have to be obtained "dynamically" by querying the AppServer objects)

> Maybe. I did consider this and am open to being persuaded to do that. But there are two reasons I didn't do that for now:
> 1. This is already a big change, and I didn't want to rewrite/change much more in this one PR.
> 2. In multi-VM setups, we won't necessarily just have multiple "edxapp" VMs. We may want to have e.g. 2 edxapp VMs, and one VM to run workers, xqueue, notifier, etc. Simply changing this from supporting one active server to supporting multiple active AppServers doesn't necessarily lay the right groundwork for that sort of approach.
> I was tentatively thinking that we might want to add a ScalableOpenEdXInstance where there are active "pools" of AppServers, e.g. an "active edxapp server pool", an "active worker server pool" etc. which each support having multiple active AppServer instances... But I haven't thought that through, and I don't think we have a plan for how that should look, do we?

Yup, I also think it will be missing a layer to handle multiple VMs & services. We can wait until we have a proper plan, I don't want to pile on this PR -- but I wanted to bring it up so we can start to think a bit about it ahead of time, especially while we are redoing the whole object structure.

Btw, another question this raises - if OpenEdXProductionInstance and ScalableOpenEdXInstance are two separate objects, does that mean we have to recreate a whole different object to scale an mono-VM instance? That's something we have to do quite often for clients. Or does that mean that, eventually, all instances will be created using the scalable class, and will just have to be able to accomodate the different types of setups?